python/feldera/rest/client.py

```diff
@@ -0,0 +1,455 @@
+from feldera.rest.connector import Connector
+from feldera.rest.attached_connector import AttachedConnector
+from feldera.rest.program import Program
+from feldera.rest.pipeline import Pipeline
+from feldera.rest._httprequests import HttpRequests
+
+
+class Client:
+    """
+    A client for the Feldera HTTP API
+
+    A client instance is needed for every Feldera API method to know the location of
+    Feldera and its permissions.
+    """
+
+    def __init__(
+            self,
+            url: str,
+            api_key: Optional[str] = None,
+            timeout: Optional[int] = None,
+    ) -> None:
+        """
+        :param url: The url to Feldera API (ex: https://try.feldera.com)
+        :param api_key: The optional API key for Feldera
+        :param timeout: (optional) The amount of time in seconds that the cient will wait for a response beforing timing
+            out.
+        """
+
+        self.config = Config(url, api_key, timeout)
+        self.http = HttpRequests(self.config)
+
+    def programs(self) -> list[Program]:
+        """
+        Get all programs
+        """
+        resp = self.http.get(
+            path="/programs",
+        )
+
+        return [
+            Program(
+                name=program.get("name"),
+                program=program.get("code"),
+                description=program.get("description"),
+            ) for program in resp
+        ]
+
+    def get_program(self, name: str, with_code: bool = False) -> Program:
+        """
+        Get a program by name
+        :param name: The name of the program
+        :param with_code: If True, the program code will be included in the response
+        """
+        resp = self.http.get(
+            path=f"/programs/{name}",
+            params={
+                "with_code": with_code,
+            }
+        )
+
+        return Program(
+            name=resp.get("name"),
+            program=resp.get("code"),
+            description=resp.get("description"),
+            status=resp.get("status"),
+            version=resp.get("version"),
+        )
+
+    def compile_program(self, program: Program):
+        """
+        Compiles a SQL program
+        :param program: The program to compile
+        """
+        body = {
+            "code": program.program,
+            "description": program.description or "",
+        }
+
+        resp = self.http.put(
+            path=f"/programs/{program.name}",
+            body=body,
+        )
+        program.id = resp.get("program_id")
+        program.version = resp.get("version")
+
+        while True:
+            status = self.get_program(program.name).status
+
+            if status == "Success":
+                break
+            elif status != "Pending" and status != "CompilingRust" and status != "CompilingSql":
+                raise RuntimeError(f"Failed program compilation with status {status}")
+
+            logging.debug("still compiling %s, waiting for 5 more seconds", program.name)

```

Comment:
Might be useful to add how long compilation has lasted already: `"Compiling program %s (so far took %.1f seconds)..."`